src/goto-instrument/contracts/contracts.cpp

```diff
@@ -686,106 +685,61 @@ void code_contractst::instrument_assign_statement(
-  goto_programt::instructionst::iterator &instruction_iterator,
-  goto_programt &program,
-  assigns_clauset &assigns_clause)
+  goto_programt::instructionst::iterator &instruction_it,
+  const irep_idt &function,
+  goto_programt &body,
+  assigns_clauset &assigns)
 {
   INVARIANT(
-    instruction_iterator->is_function_call(),
+    instruction_it->is_function_call(),
     "The first argument of instrument_call_statement should always be "
     "a function call");
 
-  irep_idt called_name;
-  if(instruction_iterator->call_function().id() == ID_dereference)
+  irep_idt called_function;
+  if(instruction_it->call_function().id() == ID_dereference)
   {
-    called_name =
+    called_function =
       to_symbol_expr(
-        to_dereference_expr(instruction_iterator->call_function()).pointer())
+        to_dereference_expr(instruction_it->call_function()).pointer())
         .get_identifier();
   }
   else
   {
-    called_name =
-      to_symbol_expr(instruction_iterator->call_function()).get_identifier();
+    called_function =
+      to_symbol_expr(instruction_it->call_function()).get_identifier();
   }
 
-  if(called_name == "malloc")
+  if(called_function == "malloc")
   {
     // malloc statments return a void pointer, which is then cast and assigned
     // to a result variable. We iterate one line forward to grab the result of
     // the malloc once it is cast.
-    instruction_iterator++;
-    if(instruction_iterator->is_assign())
+    instruction_it++;
+    if(instruction_it->is_assign())
     {
-      instrument_assign_statement(
-        instruction_iterator, program, assigns_clause);
-      const exprt &lhs = instruction_iterator->assign_lhs();
-      assigns_clause.add_to_local_write_set(dereference_exprt(lhs));
+      instrument_assign_statement(instruction_it, body, assigns);
+      const exprt &lhs = instruction_it->assign_lhs();
+      assigns.add_to_local_write_set(dereference_exprt(lhs));
     }
     return; // assume malloc edits no pre-existing memory objects.
   }
-  else if(called_name == "free")
+  else if(called_function == "free")
   {
     const exprt &lhs_dereference = dereference_exprt(
-      to_typecast_expr(instruction_iterator->call_arguments().front()).op());
-    add_containment_check(
-      program, assigns_clause, instruction_iterator, lhs_dereference);
-    assigns_clause.remove_from_local_write_set(lhs_dereference);
-    assigns_clause.remove_from_global_write_set(lhs_dereference);
+      to_typecast_expr(instruction_it->call_arguments().front()).op());
+    add_containment_check(body, assigns, instruction_it, lhs_dereference);
+    assigns.remove_from_local_write_set(lhs_dereference);
+    assigns.remove_from_global_write_set(lhs_dereference);
     return;
   }
 
   if(
-    instruction_iterator->call_lhs().is_not_nil() &&
-    instruction_iterator->call_lhs().id() == ID_symbol)
+    instruction_it->call_lhs().is_not_nil() &&
+    instruction_it->call_lhs().source_location().is_not_nil() &&
```

Comment:
You seem to be going away from "foo.is_not_nill()" and towards "!foo.is_empty()" (a move I strongly endorse) so why not here?

Reply:
We can't use `empty` here, because the return type of this methods doesn't provide us with that functionality.
`instruction_it->call_lhs()` -> `exprt &`
`instruction_it->call_lhs().source_location()` -> `const source_locationt &`
So, the correct way to check if they are "empty" is using `is_not_nil()`.